naga/src/front/atomic_upgrade.rs

```diff
@@ -0,0 +1,167 @@
+
+use crate::{GlobalVariable, Handle, Module, Type, TypeInner};
+
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum Error {
+    #[error("encountered an unsupported expression")]
+    Unsupported,
+}
+
+impl From<Error> for crate::front::spv::Error {
+    fn from(source: Error) -> Self {
+        crate::front::spv::Error::AtomicUpgradeError(source)
+    }
+}
+
+#[derive(Clone, Default)]
+struct Padding(Arc<AtomicUsize>);
+
+impl std::fmt::Display for Padding {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for _ in 0..self.0.load(std::sync::atomic::Ordering::Relaxed) {
+            f.write_str("  ")?;
+        }
+        Ok(())
+    }
+}
+
+impl Drop for Padding {
+    fn drop(&mut self) {
+        let _ = self.0.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+impl Padding {
+    fn trace(&self, msg: impl std::fmt::Display, t: impl std::fmt::Debug) {
+        format!("{msg} {t:#?}")
+            .split('\n')
+            .for_each(|ln| log::trace!("{self}{ln}"));
+    }
+
+    fn debug(&self, msg: impl std::fmt::Display, t: impl std::fmt::Debug) {
+        format!("{msg} {t:#?}")
+            .split('\n')
+            .for_each(|ln| log::debug!("{self}{ln}"));
+    }
+
+    fn inc_padding(&self) -> Padding {
+        let _ = self.0.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+        self.clone()
+    }
+}
+
+struct UpgradeState<'a> {
+    padding: Padding,
+    module: &'a mut Module,
+}
+
+impl<'a> UpgradeState<'a> {
+    fn inc_padding(&self) -> Padding {
+        self.padding.inc_padding()
+    }
+
+    /// Upgrade the type, recursing until we reach the leaves.
+    /// At the leaves, replace scalars with atomic scalars.
+    fn upgrade_type(&mut self, ty: Handle<Type>) -> Result<Handle<Type>, Error> {
+        let padding = self.inc_padding();
+        padding.trace("upgrading type: ", ty);
+
+        let r#type = self.module.types[ty].clone();
+
+        let inner = match r#type.inner.clone() {
+            TypeInner::Scalar(scalar) => {
+                log::trace!("{padding}hit the scalar leaf, replacing with an atomic");
+                TypeInner::Atomic(scalar)
+            }
+            TypeInner::Pointer { base, space } => TypeInner::Pointer {
+                base: self.upgrade_type(base)?,
+                space,
+            },
+            TypeInner::Array { base, size, stride } => TypeInner::Array {
+                base: self.upgrade_type(base)?,
+                size,
+                stride,
+            },
+            TypeInner::Struct { mut members, span } => TypeInner::Struct {
+                members: {
+                    // In the future we should have to figure out which member needs
+                    // upgrading, but for now we'll only cover the single-member case
+                    if members.len() == 1 {
+                        members[0].ty = self.upgrade_type(members[0].ty)?;
+                    }
+                    members
```

Comment:
We still have to handle the single-member case, since most types end up bundled as members in structs, it seems.